module1/homework/grayscaleImages/compression.cpp

```diff
@@ -0,0 +1,56 @@
+#include "compression.hpp"
+
+#include <cctype>
+#include <iostream>
+
+std::vector<std::pair<uint8_t, uint8_t>> compressGrayscale(
+    const std::array<std::array<uint8_t, width>, height>& bitmap) {
+    std::vector<std::pair<uint8_t, uint8_t>> compressedBitmap{};
+
+    for (const auto& el : bitmap) {
+        compressedBitmap.push_back({*begin(el), 1});
```

Comment:
How is it possible, that there is no **std::** before **begin(el)** and it works? Taking into cosideration there is no **using namespace std;**

Reply:
@nauka-programowania-MA explained during some lecture that as long as you do not have any symbol (e.g. variable or function) with name **begin** or **end** in the file, the compiler will automatically search in **std** namespace and use proper function.

Reply:
Exactly. When you use algorithm which is in namespace `std:`: there is some mechanism which will search begin and end in `std` namespace.

Reply:
It's ADL mechanism described here https://en.cppreference.com/w/cpp/language/adl